kotlin-native/runtime/src/main/kotlin/kotlin/native/internal/FloatingPointParser.kt

```diff
@@ -155,16 +249,25 @@ object FloatingPointParser {
         if (length == 0)
             throw NumberFormatException(s)
 
+        // Getting dot separator from the string (E/e)
         decimal = s.indexOf('.')
-        if (decimal > -1) {
+        s = if (decimal > -1) {
             shift = end - decimal - 1
             // Prevent e overflow, shift >= 0.
             if (e >= 0 || e - Int.MIN_VALUE > shift) {
                 e -= shift
             }
-            s = s.substring(start, decimal) + s.substring(decimal + 1, end)
+            s.substring(start, decimal) + s.substring(decimal + 1, end)
         } else {
-            s = s.substring(start, end)
+            s.substring(start, end)
+        }
+
+        // Optimal validation of characters in the string to prevent incorrect parsing.
+        // Number after an exponent were validated already.
+        s.forEach {
```

Comment:
`s` - is a body of a floating point number, for example in `500e100` - `s` = 500. But there is **no check** that it contains valid characters (digits), so we are in the situation when we can start processing of invalid string with a further logic and return Infinity in case of long string (`38` characters for `float` and `308` for `double`)